rust/connlib/tunnel/src/io/nameserver_set.rs

```diff
@@ -0,0 +1,167 @@
+                        self.socket_factory.clone(),
+                        SocketAddr::new(nameserver, crate::dns::DNS_PORT),
+                        query_firezone_dev(),
+                    ),
+                    QueryMetaData { nameserver, start },
+                )
+                .is_err()
+            {
+                tracing::debug!(%nameserver, "Failed to queue another DNS query");
+            }
+        }
+    }
+
+    pub fn fastest(&self) -> Option<IpAddr> {
+        let (_, ns) = self.nameserver_by_rtt.first_key_value()?;
+
+        Some(*ns)
+    }
+
+    pub fn poll(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        if self.queries.is_empty() {
+            return Poll::Ready(());
+        }
+
+        loop {
+            match ready!(self.queries.poll_unpin(cx)) {
+                (Ok(Ok(response)), meta) if response.header().rcode() == Rcode::NOERROR => {
+                    let rtt = meta.start.elapsed();
+
+                    tracing::debug!(nameserver = %meta.nameserver, ?rtt, ?response, "DNS query completed");
+
+                    self.nameserver_by_rtt.insert(rtt, meta.nameserver);
+                }
+                (Ok(Ok(response)), meta) => {
+                    tracing::debug!(nameserver = %meta.nameserver, ?response, "DNS query failed");
+                }
+                (Ok(Err(e)), meta) => {
+                    tracing::debug!(nameserver = %meta.nameserver, "DNS query failed: {e}");
+                }
+                (Err(_), meta) => {
+                    tracing::debug!(nameserver = %meta.nameserver, "DNS query timed out after {DNS_TIMEOUT:?}");
+                }
+            }
+
+            let Some(fastest) = self.fastest() else {
+                continue;
+            };
+
+            if self.queries.is_empty() {
+                tracing::info!(%fastest, "Evaluated fastest nameserver");
+
+                return Poll::Ready(());
+            }
+        }
+    }
+}
+
+fn query_firezone_dev() -> Message<Vec<u8>> {
+    let mut builder = MessageBuilder::new_vec().question();
+    builder.header_mut().set_random_id();
+    builder.header_mut().set_rd(true);
+    builder.header_mut().set_qr(false);
+
+    builder
+        .push(Question::new_in(FIREZONE_DEV.clone(), Rtype::A))
+        .expect("static question should always be valid");
+
+    builder.into_message()
+}
+
+#[cfg(test)]
+mod tests {
+    use std::net::Ipv4Addr;
+
+    use super::*;
+
+    #[tokio::test]
+    #[ignore = "Needs Internet"]
+    async fn can_evaluate_fastest_nameserver() {
+        let _guard = firezone_logging::test("debug");
+
+        let mut set = NameserverSet::new(
+            BTreeSet::from([
+                Ipv4Addr::new(1, 1, 1, 1).into(),
+                Ipv4Addr::new(8, 8, 8, 8).into(),
+                Ipv4Addr::new(8, 8, 4, 4).into(),
+                Ipv4Addr::new(9, 9, 9, 9).into(),
+                Ipv4Addr::new(100, 100, 100, 100).into(), // Also include an unreachable server.
```

Comment:
Would this be a possible source of a flaky test? I suppose there's the question of what happens if the admin configures a CGNAT nameserver? 🙃

Reply:
`10.10.10.10`?

Reply:
> Would this be a possible source of a flaky test? I suppose there's the question of what happens if the admin configures a CGNAT nameserver? 🙃

I am not asserting for this one failing, it is just included to ensure that we have a timeout.

Reply:
> Would this be a possible source of a flaky test? I suppose there's the question of what happens if the admin configures a CGNAT nameserver? 🙃

Admin of what? Admin of the Linux box that runs these tests? I doubt they'll be flaky in CI unless all of the above happen to have an outage at the same time.